feature/dynamodb/attributevalue/decode.go

```diff
@@ -0,0 +1,711 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeMap(avMap map[string]types.AttributeValue, v reflect.Value) error {
+	switch v.Kind() {
+	case reflect.Map:
+		t := v.Type()
+		if t.Key().Kind() != reflect.String {
+			return &UnmarshalTypeError{Value: "map string key", Type: t.Key()}
+		}
+		if v.IsNil() {
+			v.Set(reflect.MakeMap(t))
+		}
+	case reflect.Struct:
+	case reflect.Interface:
+		v.Set(reflect.MakeMap(stringInterfaceMapType))
+		v = v.Elem()
+	default:
+		return &UnmarshalTypeError{Value: "map", Type: v.Type()}
+	}
+
+	if v.Kind() == reflect.Map {
+		for k, av := range avMap {
+			key := reflect.New(v.Type().Key()).Elem()
+			key.SetString(k)
+			elem := reflect.New(v.Type().Elem()).Elem()
+			if err := d.decode(av, elem, tag{}); err != nil {
+				return err
+			}
+			v.SetMapIndex(key, elem)
+		}
+	} else if v.Kind() == reflect.Struct {
+		fields := unionStructFields(v.Type(), structFieldOptions{
+			TagKey: d.options.TagKey,
+		})
+		for k, av := range avMap {
+			if f, ok := fieldByName(fields, k); ok {
+				fv := fieldByIndex(v, f.Index, func(v *reflect.Value) bool {
+					v.Set(reflect.New(v.Type().Elem()))
+					return true // to continue the loop.
+				})
+				if err := d.decode(av, fv, f.tag); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeNull(v reflect.Value) error {
+	if v.IsValid() && v.CanSet() {
+		v.Set(reflect.Zero(v.Type()))
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeString(s string, v reflect.Value, fieldTag tag) error {
+	if fieldTag.AsString {
+		return d.decodeNumber(s, v, fieldTag)
+	}
+
+	// To maintain backwards compatibility with ConvertFrom family of methods which
+	// converted strings to time.Time structs
+	if v.Type().ConvertibleTo(timeType) {
+		t, err := time.Parse(time.RFC3339, s)
+		if err != nil {
+			return err
+		}
+		v.Set(reflect.ValueOf(t).Convert(v.Type()))
+		return nil
+	}
+
+	switch v.Kind() {
+	case reflect.String:
+		v.SetString(s)
+	case reflect.Interface:
+		// Ensure type aliasing is handled properly
+		v.Set(reflect.ValueOf(s).Convert(v.Type()))
+	default:
+		return &UnmarshalTypeError{Value: "string", Type: v.Type()}
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeStringSet(ss []string, v reflect.Value) error {
+	switch v.Kind() {
+	case reflect.Slice:
+		// Make room for the slice elements if needed
+		if v.IsNil() || v.Cap() < len(ss) {
+			v.Set(reflect.MakeSlice(v.Type(), 0, len(ss)))
+		}
+	case reflect.Array:
+		// Limited to capacity of existing array.
+	case reflect.Interface:
+		set := make([]string, len(ss))
+		for i, s := range ss {
+			if err := d.decodeString(s, reflect.ValueOf(&set[i]).Elem(), tag{}); err != nil {
+				return err
+			}
+		}
+		v.Set(reflect.ValueOf(set))
+		return nil
+	default:
+		return &UnmarshalTypeError{Value: "string set", Type: v.Type()}
+	}
+
+	for i := 0; i < v.Cap() && i < len(ss); i++ {
+		v.SetLen(i + 1)
```

Comment:
V1 had a condition here explicitly if the type was `reflect.Array` determined by the switch statement above. https://github.com/aws/aws-sdk-go/blob/d7b5372b5b7583d5a30a3d7440d9d684aba581a9/service/dynamodb/dynamodbattribute/decode.go#L588 From the docs v.SetLen will panic if v is not a slice.

Reply:
reference original issue from v1, https://github.com/aws/aws-sdk-go/issues/1594